hal/src/thumbv7em/pukcc.rs

```diff
@@ -0,0 +1,579 @@
+            ),+
+        ) =>
+        {
+            {
+            (&[])
+                .iter()
+            $(
+                .chain($data.iter().rev()) // > Little endian format! <
+            )+
+            .zip($crypto_ram.iter_mut())
+            .for_each(|(data_iter, cr_iter)| *cr_iter = *data_iter);
+
+            let mut _offset = 0;
+            $(
+                $name = &$crypto_ram[_offset.._offset + $data.len()];
+                _offset += $data.len();
+            )+
+            }
+        }
+}
+
+/// Struct representing a PUKCC peripheral.
+///
+/// It provides an access to cryptograhic algorithms in a safe, high-level
+/// manner
+pub struct Pukcc {
+    __: (),
+}
+
+impl Pukcc {
+    /// Constructor.
+    ///
+    /// Waits for a CryptoRAM readiness, enables a synchronous PUKCC clock and
+    /// performs a self test. In case a self test fails it returns an error
+    pub fn enable(mclk: &mut MCLK) -> Result<Self, SelfTestFailure> {
+        // TODO: Alleged RNG initialization. Passthrough constructor?
+        unsafe {
+            c_abi::wait_for_crypto_ram_clear_process();
+        }
+        mclk.ahbmask.modify(|_, w| w.pukcc_().set_bit());
+        let pukcc = Self { __: () };
+        pukcc.self_test().map(|_| pukcc)
+    }
+
+    /// Self test service.
+    ///
+    /// Clears up a CryptoRAM and does the checksum. If a checksum and a version
+    /// matches one defined in a HAL, it means that a self test passed
+    /// successfully.
+    ///
+    /// While using a high-level API, user should not need to use this service
+    /// explicitly.
+    pub fn self_test(&self) -> Result<(), SelfTestFailure> {
+        const PUKCL_VERSION: u4 = 0x04070100;
+        const CHECKNUM_1: u4 = 0x6E70DDD2;
+        const CHECKNUM_2: u4 = 0x25C8D64F;
+        let mut pukcl_params = c_abi::PukclParams::default();
+        unsafe {
+            c_abi::SelfTest::call(&mut pukcl_params);
+        }
+        let header = pukcl_params.header;
+        let service_params = unsafe { pukcl_params.params.SelfTest };
+        match header.u2Status.into() {
+            PukclReturnCode::Ok => {}
+            _ => return Err(SelfTestFailure(service_params)),
+        };
+        if service_params.u4Version != PUKCL_VERSION {
+            return Err(SelfTestFailure(service_params));
+        }
+        if service_params.u4CheckNum1 != CHECKNUM_1 {
+            return Err(SelfTestFailure(service_params));
+        }
+        if service_params.u4CheckNum2 != CHECKNUM_2 {
+            return Err(SelfTestFailure(service_params));
+        }
+
+        Ok(())
+    }
+
+    /// Service generating an ECDSA signature.
+    ///
+    /// GF(p) service. GF(2^n) variant is not implemented -- use low-level API.
+    ///
+    /// Input parameters:
+    /// - `hash`: `&[u8]` of length [`Curve::SCALAR_LENGTH`]
+    ///     - Hash of a message that is supposed to be signed.
+    /// - `private_key`: `&[u8]` of length [`Curve::SCALAR_LENGTH`]
```

Comment:
Aside from the length property, is it possible to pass in a private key that is invalid? If so, can we create a new type that represents a private key with valid parameters, and only have safe ways of creating that type?

Reply:
I don't know. I'm not even sure if these lengths for slices are correct as I relied on a documentation and its max lengths fields from these tables. My cryptographical expertise is also limited, to say the least. Plus documentation is full of typos and errors tbf. Private key is just a [u8;32] for this implemented curve. I don't think we can prevent the user from passing slice of zeroes for example :shrug: